File: 0938. Range Sum of BST.js
// INPUT: 10
//       /  \
//      5    15
//     / \    \
//    3   7    18
// min: 7, max: 15
// OUTPUT: 32

const rangeSumBST = (root, min, max) => {
  const traverse = (node) => {
    if (node === null) return 0;
    if (node.val > max) return traverse(node.left);
    if (node.val < min) return traverse(node.right);

    return traverse(node.left) + node.val + traverse(node.right);
  };

  return traverse(root);
};

console.log(
  rangeSumBST(
    {
      val: 10,
      left: {
        val: 5,
        left: { val: 3, left: null, right: null },
        right: { val: 7, left: null, right: null },
      },
      right: {
        val: 15,
        left: null,
        right: { val: 18, left: null, right: null },
      },
    },
    7,
    15
  )
);

File: python/0938. Range Sum of BST.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rangeSumBST(self, root: TreeNode, low: int,
                    high: int) -> int:
        if root is None:
            return 0

        if root.val < low:
            sum_left = 0
        else:
            sum_left = self.rangeSumBST(root.left, low, high)

        if root.val > high:
            sum_right = 0
        else:
            sum_right = self.rangeSumBST(root.right, low, high)

        if low <= root.val <= high:
            sum_curr = root.val
        else:
            sum_curr = 0

        return sum_left + sum_curr + sum_right
